### PR TITLE
Implement KORE sort inference

### PR DESCRIPTION
**Pull request by tothtamas28**

### CODE REVIEW

**Comment by tothtamas28** [comment]
Here's an example KAST term from an actual proof execution (from https://github.com/runtimeverification/evm-semantics/pull/1594) where the simple sort inference algorithm fails due to non-unique quantifier names returned by the Haskell backend:

[kast.json.txt](https://github.com/runtimeverification/pyk/files/10911502/kast.json.txt)

The modified algorithm (`kast_to_kore_2`) converts the term to KORE without problem.

```python3
from pathlib import Path

from pyk.kast.inner import KInner
from pyk.kore.kompiled import KompiledKore
from pyk.konvert import kast_to_kore_2

kast = KInner.from_json(Path('kast.json.txt').read_text())
kompiled_kore = KompiledKore('../evm-semantics/.build/usr/lib/kevm/haskell/')
kore = kast_to_kore_2(kompiled_kore, kast)
```

**Comment by tothtamas28** [comment]
## Performance Measurements

### Conversion with KAST-based sort inference

Profiling script: [profile_kast.py.txt](https://github.com/runtimeverification/pyk/files/10919378/profile_kast.py.txt)

Single run:

```
$ poetry run python3 profile_kast.py
```

```
         256626 function calls (238936 primitive calls) in 0.129 seconds

   Ordered by: cumulative time
   List reduced from 203 to 20 due to restriction <20>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.142    0.142 /home/ttoth/git/pyk/src/pyk/konvert.py:49(kast_to_kore)
        1    0.000    0.000    0.071    0.071 /home/ttoth/git/pyk/src/pyk/kast/outer.py:1128(sort_vars)
        1    0.000    0.000    0.070    0.070 /home/ttoth/git/pyk/src/pyk/kast/outer.py:1087(sort_vars_subst)
   1389/5    0.001    0.000    0.070    0.014 /home/ttoth/git/pyk/src/pyk/kast/inner.py:666(bottom_up)
    911/5    0.000    0.000    0.070    0.014 /home/ttoth/git/pyk/src/pyk/kast/inner.py:443(map_inner)
    911/5    0.001    0.000    0.070    0.014 /home/ttoth/git/pyk/src/pyk/kast/inner.py:438(let)
    913/5    0.002    0.000    0.070    0.014 /home/ttoth/git/pyk/src/pyk/kast/inner.py:395(__init__)
  2285/15    0.000    0.000    0.070    0.005 /home/ttoth/git/pyk/src/pyk/kast/inner.py:444(<genexpr>)
  1384/10    0.000    0.000    0.070    0.007 /home/ttoth/git/pyk/src/pyk/kast/inner.py:667(<lambda>)
        2    0.000    0.000    0.066    0.033 /home/ttoth/git/pyk/src/pyk/kast/inner.py:691(collect)
      558    0.000    0.000    0.064    0.000 /home/ttoth/git/pyk/src/pyk/kast/inner.py:692(f)
      279    0.000    0.000    0.064    0.000 /home/ttoth/git/pyk/src/pyk/kast/outer.py:1091(_sort_contexts)
      183    0.000    0.000    0.063    0.000 /home/ttoth/git/pyk/src/pyk/kast/outer.py:1033(production_for_klabel)
      134    0.059    0.000    0.059    0.000 /home/ttoth/git/pyk/src/pyk/kast/outer.py:1035(<listcomp>)
    279/1    0.000    0.000    0.044    0.044 /home/ttoth/git/pyk/src/pyk/kore/kompiled.py:97(add_injections)
    420/4    0.001    0.000    0.043    0.011 /home/ttoth/.pyenv/versions/3.10.10/lib/python3.10/functools.py:961(__get__)
  352/179    0.000    0.000    0.042    0.000 /home/ttoth/git/pyk/src/pyk/kore/syntax.py:413(__init__)
      6/1    0.000    0.000    0.040    0.040 /home/ttoth/git/pyk/src/pyk/kore/syntax.py:703(let_patterns)
    557/3    0.000    0.000    0.040    0.013 /home/ttoth/git/pyk/src/pyk/kore/kompiled.py:102(<genexpr>)
    171/7    0.000    0.000    0.040    0.006 /home/ttoth/git/pyk/src/pyk/kore/syntax.py:431(let_patterns)
```

1000 runs:

```
$ poetry run python3 profile_kast.py 1000
```

```
         99176608 function calls (82080014 primitive calls) in 27.920 seconds

   Ordered by: cumulative time
   List reduced from 204 to 20 due to restriction <20>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     1000    0.028    0.000   32.968    0.033 /home/ttoth/git/pyk/src/pyk/konvert.py:49(kast_to_kore)
278000/1000    0.313    0.000   21.130    0.021 /home/ttoth/git/pyk/src/pyk/konvert.py:72(_kast_to_kore)
183000/1000    0.065    0.000   20.112    0.020 /home/ttoth/git/pyk/src/pyk/konvert.py:135(_kapply_to_kore)
183000/1000    0.346    0.000   20.111    0.020 /home/ttoth/git/pyk/src/pyk/konvert.py:153(_kapply_to_pattern)
459000/3000    0.095    0.000   20.087    0.007 /home/ttoth/git/pyk/src/pyk/konvert.py:157(<genexpr>)
7007000/278000    5.464    0.000   12.126    0.000 /home/ttoth/.pyenv/versions/3.10.10/lib/python3.10/dataclasses.py:232(wrapper)
352000/179000    0.437    0.000    9.208    0.000 /home/ttoth/git/pyk/src/pyk/kore/syntax.py:413(__init__)
   534002    0.216    0.000    7.013    0.000 /home/ttoth/git/pyk/src/pyk/kore/lexer.py:271(id)
   534013    3.755    0.000    6.797    0.000 /home/ttoth/git/pyk/src/pyk/kore/lexer.py:390(_id_text)
279000/1000    0.373    0.000    6.591    0.007 /home/ttoth/git/pyk/src/pyk/kore/kompiled.py:97(add_injections)
6000/1000    0.005    0.000    6.556    0.007 /home/ttoth/git/pyk/src/pyk/kore/syntax.py:703(let_patterns)
557000/3000    0.147    0.000    6.554    0.002 /home/ttoth/git/pyk/src/pyk/kore/kompiled.py:102(<genexpr>)
171000/7000    0.066    0.000    6.243    0.001 /home/ttoth/git/pyk/src/pyk/kore/syntax.py:431(let_patterns)
171000/7000    0.146    0.000    6.240    0.001 /home/ttoth/git/pyk/src/pyk/kore/syntax.py:419(let)
   352011    0.247    0.000    5.706    0.000 /home/ttoth/git/pyk/src/pyk/kore/syntax.py:35(check_symbol_id)
     1000    0.001    0.000    5.224    0.005 /home/ttoth/git/pyk/src/pyk/kore/syntax.py:1172(let_patterns)
   352011    0.100    0.000    5.083    0.000 /home/ttoth/git/pyk/src/pyk/kore/lexer.py:277(symbol_id)
1389000/5000    0.720    0.000    4.910    0.001 /home/ttoth/git/pyk/src/pyk/kast/inner.py:666(bottom_up)
911000/5000    0.467    0.000    4.903    0.001 /home/ttoth/git/pyk/src/pyk/kast/inner.py:443(map_inner)
911000/5000    0.542    0.000    4.900    0.001 /home/ttoth/git/pyk/src/pyk/kast/inner.py:438(let)
```

### Conversion with KORE-based sort inference

Profiling script: [profile_kore.py.txt](https://github.com/runtimeverification/pyk/files/10919381/profile_kore.py.txt)

Single run:

```
$ poetry run python3 profile_kore.py
```

```
         283028 function calls (271626 primitive calls) in 0.064 seconds

   Ordered by: cumulative time
   List reduced from 162 to 20 due to restriction <20>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
        1    0.000    0.000    0.070    0.070 /home/ttoth/git/pyk/src/pyk/konvert.py:63(kast_to_kore_2)
        1    0.000    0.000    0.042    0.042 /home/ttoth/git/pyk/src/pyk/kore/kompiled.py:116(strengthen_sorts)
     13/2    0.000    0.000    0.033    0.017 /home/ttoth/.pyenv/versions/3.10.10/lib/python3.10/functools.py:961(__get__)
        1    0.000    0.000    0.031    0.031 /home/ttoth/git/pyk/src/pyk/kore/kompiled.py:204(map)
        1    0.000    0.000    0.031    0.031 /home/ttoth/git/pyk/src/pyk/kore/kompiled.py:206(<dictcomp>)
       44    0.000    0.000    0.031    0.001 /home/ttoth/git/pyk/src/pyk/kore/kompiled.py:93(meet_all_sorts)
      101    0.000    0.000    0.031    0.000 /home/ttoth/git/pyk/src/pyk/kore/kompiled.py:63(is_subsort)
       45    0.000    0.000    0.031    0.001 {built-in method _functools.reduce}
       46    0.000    0.000    0.031    0.001 /home/ttoth/git/pyk/src/pyk/kore/kompiled.py:75(meet_sorts)
        1    0.005    0.005    0.030    0.030 /home/ttoth/git/pyk/src/pyk/kore/kompiled.py:39(_subsort_table)
    278/1    0.000    0.000    0.021    0.021 /home/ttoth/git/pyk/src/pyk/konvert.py:72(_kast_to_kore)
    183/1    0.000    0.000    0.020    0.020 /home/ttoth/git/pyk/src/pyk/konvert.py:135(_kapply_to_kore)
    183/1    0.000    0.000    0.020    0.020 /home/ttoth/git/pyk/src/pyk/konvert.py:153(_kapply_to_pattern)
    459/3    0.000    0.000    0.020    0.007 /home/ttoth/git/pyk/src/pyk/konvert.py:157(<genexpr>)
      752    0.000    0.000    0.016    0.000 /home/ttoth/git/pyk/src/pyk/kore/kompiled.py:44(<genexpr>)
      752    0.005    0.000    0.016    0.000 /home/ttoth/git/pyk/src/pyk/kore/kompiled.py:43(<genexpr>)
  523/186    0.001    0.000    0.014    0.000 /home/ttoth/git/pyk/src/pyk/kore/syntax.py:413(__init__)
 7007/278    0.005    0.000    0.012    0.000 /home/ttoth/.pyenv/versions/3.10.10/lib/python3.10/dataclasses.py:232(wrapper)
     12/2    0.000    0.000    0.012    0.006 /home/ttoth/git/pyk/src/pyk/kore/syntax.py:703(let_patterns)
    30427    0.007    0.000    0.011    0.000 /home/ttoth/git/pyk/src/pyk/kore/kompiled.py:42(<genexpr>)
```

1000 runs:

```
$ poetry run python3 profile_kore.py 1000
```

```
         112645455 function calls (101414284 primitive calls) in 28.943 seconds

   Ordered by: cumulative time
   List reduced from 162 to 20 due to restriction <20>

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
     1000    0.032    0.000   33.648    0.034 /home/ttoth/git/pyk/src/pyk/konvert.py:63(kast_to_kore_2)
278000/1000    0.302    0.000   19.712    0.020 /home/ttoth/git/pyk/src/pyk/konvert.py:72(_kast_to_kore)
183000/1000    0.063    0.000   18.752    0.019 /home/ttoth/git/pyk/src/pyk/konvert.py:135(_kapply_to_kore)
183000/1000    0.341    0.000   18.751    0.019 /home/ttoth/git/pyk/src/pyk/konvert.py:153(_kapply_to_pattern)
459000/3000    0.092    0.000   18.729    0.006 /home/ttoth/git/pyk/src/pyk/konvert.py:157(<genexpr>)
523000/186000    0.650    0.000   13.382    0.000 /home/ttoth/git/pyk/src/pyk/kore/syntax.py:413(__init__)
7007000/278000    5.223    0.000   11.538    0.000 /home/ttoth/.pyenv/versions/3.10.10/lib/python3.10/dataclasses.py:232(wrapper)
12000/2000    0.010    0.000   10.867    0.005 /home/ttoth/git/pyk/src/pyk/kore/syntax.py:703(let_patterns)
342000/14000    0.122    0.000   10.471    0.001 /home/ttoth/git/pyk/src/pyk/kore/syntax.py:431(let_patterns)
342000/14000    0.234    0.000   10.466    0.001 /home/ttoth/git/pyk/src/pyk/kore/syntax.py:419(let)
   929002    0.342    0.000    9.856    0.000 /home/ttoth/git/pyk/src/pyk/kore/lexer.py:271(id)
   929013    4.829    0.000    9.514    0.000 /home/ttoth/git/pyk/src/pyk/kore/lexer.py:390(_id_text)
     2000    0.002    0.000    8.772    0.004 /home/ttoth/git/pyk/src/pyk/kore/syntax.py:1172(let_patterns)
   523011    0.372    0.000    8.389    0.000 /home/ttoth/git/pyk/src/pyk/kore/syntax.py:35(check_symbol_id)
     1000    0.003    0.000    7.512    0.008 /home/ttoth/git/pyk/src/pyk/kore/kompiled.py:116(strengthen_sorts)
   523011    0.142    0.000    7.487    0.000 /home/ttoth/git/pyk/src/pyk/kore/lexer.py:277(symbol_id)
279000/1000    0.362    0.000    6.386    0.006 /home/ttoth/git/pyk/src/pyk/kore/kompiled.py:97(add_injections)
557000/3000    0.142    0.000    6.353    0.002 /home/ttoth/git/pyk/src/pyk/kore/kompiled.py:102(<genexpr>)
279000/1000    0.166    0.000    4.514    0.005 /home/ttoth/git/pyk/src/pyk/kore/kompiled.py:137(rewrite)
232000/1000    0.165    0.000    4.512    0.005 /home/ttoth/git/pyk/src/pyk/kore/syntax.py:278(map_patterns)
```

**Comment by ehildenb** [comment]
My reading of the profiles here is summarized as follows:

- Not really any performance difference.
- In both cases, the sort inference (at kast level or at kore level) is tiny compared to other components of the algorithm.

**Comment by ehildenb** [comment]
Ok, I'm convinced, the code change is nice and compact, looks easy enough to go back. Let's just have the default implementation of `kast_to_kore` go through the new code path instead.

File: src/pyk/konvert.py [hunk 1]
```diff
@@ -60,6 +60,22 @@ def kast_to_kore(
...
 
+def kast_to_kore_2(
+    kast_defn: KDefinition,
+    kompiled_kore: KompiledKore,
+    kast: KInner,
+    sort: Optional[KSort] = None,
+) -> Pattern:
+    if sort is None:
+        sort = K
+    kast = kast_defn.add_cell_map_items(kast)
```
Comment (ehildenb) [inline src/pyk/konvert.py:71]:
```suggestion
    # TODO: should not need to mangle kast CellMap items like this.
    kast = kast_defn.add_cell_map_items(kast)
```

**Pull request closed without merging**